test/integration/multicluster/multicluster-traffic/mc_traffic_test.go

```diff
@@ -0,0 +1,369 @@
+	// Before starting, get source context
+	contexts = TestHelper.GetMulticlusterContexts()
+	targetCtx = contexts["tgt"]
+	// Then, re-build clientset with context of target cluster instead of kube
+	// context inferred from environment.
+	if err := TestHelper.SwitchContext(targetCtx); err != nil {
+		out := fmt.Sprintf("Error running test: failed to switch Kubernetes client to context [%s]: %s\n", targetCtx, err)
+		os.Stderr.Write([]byte(out))
+		os.Exit(1)
+	}
+	// Block until viz deploy is running successfully in target cluster.
+	TestHelper.WaitUntilDeployReady(testutil.LinkerdVizDeployReplicas)
+	os.Exit(m.Run())
+}
+
+// TestGateways tests the `linkerd multicluster gateways` command by installing
+// three emojivoto services in target cluster and asserting the output against
+// the source cluster.
+func TestGateways(t *testing.T) {
+	if err := TestHelper.SwitchContext(contexts["tgt"]); err != nil {
+		testutil.AnnotatedFatalf(t, "failed to rebuild helper clientset with new context", "failed to rebuild helper clientset with new context [%s]: %v", contexts["tgt"], err)
+	}
+
+	t.Run("install resources in target cluster", func(t *testing.T) {
+		// Create namespace in source cluster
+		out, err := TestHelper.KubectlWithContext("", contexts["src"], "create", "namespace", "linkerd-nginx-gateway-deploy")
+		if err != nil {
+			testutil.AnnotatedFatalf(t, "failed to create namespace", "failed to create namespace 'linkerd-nginx-gateway-deploy': %s\n%s", err, out)
+		}
+
+		out, err = TestHelper.KubectlApplyWithContext("", contexts["tgt"], "-n", "linkerd-nginx-gateway-deploy", "-f", "testdata/nginx-gateway-deploy.yaml")
+		if err != nil {
+			testutil.AnnotatedFatalf(t, "failed to install nginx deploy", "failed to install nginx deploy: %s\n%s", err, out)
+		}
+
+		// Wait for workloads to spin up in target cluster. These workloads will have
+		// their services mirrored in the source cluster. The test will check whether
+		// the gateway keeps track of the mirror services.
+		tgtWorkloadRollouts := map[string]testutil.DeploySpec{
+			"nginx-deploy": {Namespace: "linkerd-nginx-gateway-deploy", Replicas: 1},
+		}
+		TestHelper.WaitRolloutWithContext(t, tgtWorkloadRollouts, contexts["tgt"])
+	})
+
+	timeout := time.Minute
+	err := TestHelper.RetryFor(timeout, func() error {
+		out, err := TestHelper.LinkerdRun("--context="+contexts["src"], "multicluster", "gateways")
+		if err != nil {
+			return err
+		}
+		rows := strings.Split(out, "\n")
+		if len(rows) < 2 {
+			return errors.New("response is empty")
+		}
+		fields := strings.Fields(rows[1])
+		if len(fields) < 6 {
+			return fmt.Errorf("unexpected number of columns: %d", len(fields))
+		}
+		if fields[0] != "target" {
+			return fmt.Errorf("unexpected target cluster name: %s", fields[0])
+		}
+		if fields[1] != "True" {
+			return errors.New("target cluster is not alive")
+		}
+		if fields[2] != "1" {
+			return fmt.Errorf("invalid NUM_SVC: %s", fields[2])
+		}
```

Comment:
This makes me think it'd be helpful if `linkerd multicluster gateways` had a `-o json` flag.

Reply:
Yeah, I think it'd be great to have json format for it. I think generally switching to json output in tests should make things easier. Even for checks. It'd be much easier to check what we need from a json object as opposed to using templates and golden files.

Reply:
#8285